doc/primesieve.1
'\" t
.\"     Title: primesieve
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 02/17/2024
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "PRIMESIEVE" "1" "02/17/2024" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
primesieve \- generate prime numbers
.SH "SYNOPSIS"
.sp
\fBprimesieve\fR [\fISTART\fR] \fISTOP\fR [\fIOPTION\fR]\&...
.SH "DESCRIPTION"
.sp
Generate the prime numbers and/or prime k\-tuplets inside [\fISTART\fR, \fISTOP\fR] (< 2^64) using the segmented sieve of Eratosthenes\&. primesieve includes a number of extensions to the sieve of Eratosthenes which significantly improve performance: multiples of small primes are pre\-sieved, it uses wheel factorization to skip multiples with small prime factors and it uses the bucket sieve algorithm which improves cache efficiency when sieving > 2^32\&. primesieve is also multi\-threaded, it uses all available CPU cores by default for counting primes and for finding the nth prime\&.
.sp
The segmented sieve of Eratosthenes has a runtime complexity of O(n log log n) operations and it uses O(n^(1/2)) bits of memory\&. More specifically primesieve uses 8 bytes per sieving prime, hence its memory usage can be approximated by PrimePi(n^(1/2)) * 8 bytes (per thread)\&.
.SH "OPTIONS"
.PP
\fB\-c\fR[\fINUM+\fR], \fB\-\-count\fR[=\fINUM+\fR]
.RS 4
Count primes and/or prime k\-tuplets, 1 <=
\fINUM\fR
<= 6\&. Count primes:
\fB\-c\fR
or
\fB\-\-count\fR, count twin primes:
\fB\-c2\fR
or
\fB\-\-count=2\fR, count prime triplets:
\fB\-c3\fR
or
\fB\-\-count=3\fR, \&... You can also count primes and prime k\-tuplets at the same time, e\&.g\&.
\fB\-c123\fR
counts primes, twin primes and prime triplets\&.
.RE
.PP
\fB\-\-cpu\-info\fR
.RS 4
Print CPU information: CPU name, frequency, number of cores, cache sizes, \&...
.RE
.PP
\fB\-d, \-\-dist\fR=\fIDIST\fR
.RS 4
Sieve the interval [\fISTART\fR,
\fISTART\fR
+
\fIDIST\fR]\&.
.RE
.PP
\fB\-h, \-\-help\fR
.RS 4
Print this help menu\&.
.RE
.PP
\fB\-n, \-\-nth\-prime\fR
.RS 4
Find the nth prime, e\&.g\&. 100
\fB\-n\fR
finds the 100th prime\&. If 2 numbers
\fIN\fR
\fISTART\fR
are provided finds the nth prime >
\fISTART\fR, e\&.g\&. 2 100
\fB\-n\fR
finds the 2nd prime > 100\&.
.RE
.PP
\fB\-\-no\-status\fR
.RS 4
Turn off the progressing status\&.
.RE
.PP
\fB\-p\fR[\fINUM\fR], \fB\-\-print\fR[=\fINUM\fR]
.RS 4
Print primes or prime k\-tuplets, 1 <=
\fINUM\fR
<= 6\&. Print primes:
\fB\-p\fR, print twin primes:
\fB\-p2\fR, print prime triplets:
\fB\-p3\fR, \&...
.RE
.PP
\fB\-q, \-\-quiet\fR
.RS 4
Quiet mode, prints less output\&.
.RE
.PP
\fB\-R, \-\-RiemannR\fR
.RS 4
Approximate PrimePi(x) using the Riemann R function: R(x)\&.
.RE
.PP
\fB\-\-RiemannR\-inverse\fR
.RS 4
Approximate the nth prime using the inverse Riemann R function: R^\-1(x)\&.
.RE
.PP
\fB\-s, \-\-size\fR=\fISIZE\fR
.RS 4
Set the size of the sieve array in KiB, 16 <=
\fISIZE\fR
<= 8192\&. By default primesieve uses a sieve size that matches your CPU\(cqs L1 cache size (per core) or is slightly smaller than your CPU\(cqs L2 cache size\&. This setting is crucial for performance, on exotic CPUs primesieve sometimes fails to determine the CPU\(cqs cache sizes which usually causes a big slowdown\&. In this case you can get a significant speedup by manually setting the sieve size to your CPU\(cqs L1 or L2 cache size (per core)\&.
.RE
.PP
\fB\-S, \-\-stress\-test\fR[=\fIMODE\fR]
.RS 4
Run a stress test\&. The
\fIMODE\fR
can be either CPU (default) or RAM\&. The CPU
\fIMODE\fR
uses little memory (< 5 MiB per thread) and puts the highest load on the CPU\&. The RAM
\fIMODE\fR
on the other hand uses much more memory than the CPU
\fIMODE\fR
(each thread uses about 1\&.16 GiB), but the CPU usually won\(cqt get as hot as in the CPU
\fIMODE\fR\&. Stress testing keeps on running until either a miscalculation occurs (due to a hardware issue) or the timeout expires\&. The default timeout is 24 hours, the timeout can be changed using the
\fB\-\-timeout=SECS\fR
option\&. By default the stress test uses a number of threads that matches the number of CPU cores, the number of threads can be changed using the
\fB\-\-threads=NUM\fR
option\&.
.RE
.PP
\fB\-\-test\fR
.RS 4
Run various correctness tests (< 1 minute)\&.
.RE
.PP
\fB\-t, \-\-threads\fR=\fINUM\fR
.RS 4
Set the number of threads, 1 <=
\fINUM\fR
<= CPU cores\&. By default primesieve uses all available CPU cores for counting primes and for finding the nth prime\&.
.RE
.PP
\fB\-\-time\fR
.RS 4
Print the time elapsed in seconds\&.
.RE
.PP
\fB\-\-timeout\fR=\fISECS\fR
.RS 4
Set the stress test timeout in seconds\&. Units of time for seconds, minutes, hours, days and years are also supported with the suffix s, m, h, d or y\&. E\&.g\&.
\fB\-\-timeout 10m\fR
sets a timeout of 10 minutes\&. The default stress test timeout is 24 hours\&.
.RE
.PP
\fB\-v, \-\-version\fR
.RS 4
Print version and license information\&.
.RE
.SH "EXAMPLES"
.PP
\fBprimesieve 1000\fR
.RS 4
Count the primes <= 1000\&.
.RE
.PP
\fBprimesieve 1e6 \-\-print\fR
.RS 4
Print the primes <= 10^6\&.
.RE
.PP
\fBprimesieve 1e6 \-\-print > primes\&.txt\fR
.RS 4
Store the primes <= 10^6 in a text file\&.
.RE
.PP
\fBprimesieve 2^32 \-\-print=2\fR
.RS 4
Print the twin primes <= 2^32\&.
.RE
.PP
\fBprimesieve 1e16 \-\-dist=1e10 \-\-threads=1\fR
.RS 4
Count the primes inside [10^16, 10^16 + 10^10] using a single thread\&.
.RE
.SH "HOMEPAGE"
.sp
https://github\&.com/kimwalisch/primesieve
.SH "AUTHOR"
.sp
Kim Walisch <kim\&.walisch@gmail\&.com>


CHANGES
Changes in version 4.2, 03/03/2012
==================================

1. libprimesieve generates (callback) primes up to 5% faster on
   little-endian CPUs due to improved endiansafe_cast.h.
2. The Makefile now supports the Solaris OS.
3. Fixed a bug in the primesieve GUI code (read BUGS).
4. Updated ExpressionParser.h (src/apps/*) to version 2.5.
5. Added nth prime example program (./examples).
6. Lots of refactoring to make the code easier to understand.

Changes in version 4.1, 12/01/2012
==================================

1. New PrimeSieveCallback interface class that simplifies prime
   number generation with classes, see doc/USAGE_EXAMPLES.
2. The primesieve console (terminal) application now supports
   GNU-style long options e.g. --size=256.
3. Fixed a bug (for sizeof(int) > 4) in the bit population count
   algorithm, read BUGS file.
4. Updated files: README, INSTALL, doc/USAGE_EXAMPLES.
5. The Makefile now uses the system's default C++ compiler instead
   of GNU g++ which makes it more portable.
6. The primesieve_error class has been moved into its own file
   src/soe/primesieve_error.h (previously PrimeSieve.h).

   API changes
   -----------

   The following two overly complex PrimeSieve methods have been
   removed:

   void PrimeSieve::generatePrimes(uint32_t start, uint32_t stop, void (*)(uint32_t, void*), void*);
   void PrimeSieve::generatePrimes(uint64_t start, uint64_t stop, void (*)(uint64_t, void*), void*);

   They are replaced by the following two new methods that greatly
   simplify prime number generation with classes, please refer
   to doc/USAGE_EXAMPLES for more information.

   void PrimeSieve::generatePrimes(uint32_t start, uint32_t stop, PrimeSieveCallback<uint32_t>*);
   void PrimeSieve::generatePrimes(uint32_t start, uint32_t stop, PrimeSieveCallback<uint64_t>*);

Changes in version 4.0, 18/10/2012
==================================

primesieve 4.0 is a major new release. A lot of work has been put into
documentation and improving primesieve as a library. The API of
primesieve 4.0 is backward incompatible with primesieve 3.*, the few
changes are documented further down.

 1. Added support for OpenMP 2.*, useful for compilers that do
    not support OpenMP >= 3.0, e.g. MSVC, Apple g++.
 2. New examples directory with 12 simple example programs.
 3. New doc/API file that lists the public member functions of the
    PrimeSieve and ParallelPrimeSieve C++ classes.
 4. The Makefile is now POSIX compatible, it works with any POSIX
    shell e.g. sh, bash, ash, ksh, zsh, ...
 5. The Makefile now supports MinGW (with MSYS) and Cygwin.
 6. Fixed a shared libprimesieve bug (read doc/BUGS).
 7. Up to 10 percent faster prime number generation due to new
    internal 64-bit getNextPrime() (previously 32-bit).
 8. New unsynchronized ParallelPrimeSieve::generatePrimes() method
    that calls back primes in parallel (read USAGE_EXAMPLES).
 9. Faster thread synchronization in ParallelPrimeSieve, replaced slow
    OpenMP critical directive with faster omp_test_lock().
10. Optimized prime k-tuplet (twin primes, ...) counting for
    out-of-order CPUs (src/soe/PrimeNumberFinder.cpp).
11. New primesieve_error() exception used for all exceptions within
    PrimeSieve and ParallelPrimeSieve (read USAGE_EXAMPLES).
12. New pre-sieve code (src/soe/PreSieve.cpp).
13. Ported the primesieve GUI application from Qt 4 to Qt 5.

    API changes
    -----------

    The PrimeSieve 3.* count methods:

    uint64_t getPrimeCount(uint64_t start, uint64_t stop);
    uint64_t getTwinCount (uint64_t start, uint64_t stop);
    ...

    Have been renamed to:

    uint64_t countPrimes     (uint64_t start, uint64_t stop);
    uint64_t countTwins      (uint64_t start, uint64_t stop);
    uint64_t countTriplets   (uint64_t start, uint64_t stop);
    uint64_t countQuadruplets(uint64_t start, uint64_t stop);
    uint64_t countQuintuplets(uint64_t start, uint64_t stop);
    uint64_t countSextuplets (uint64_t start, uint64_t stop);
    uint64_t countSeptuplets (uint64_t start, uint64_t stop);

Changes in version 3.8, 13/07/2012
==================================

 1. Improved OOP design of WheelFactorization.h.
 2. Minor speed up for big sieving primes ~2% (src/soe/EratBig.cpp),
    reduced the number of operations in the main sieving loop.
 3. Minor speed up for small sieving primes ~3% (src/soe/EraSmall.cpp),
    new inner sieving loop without instruction dependencies that uses
    only 12 asm instructions (previously 16).
 4. Improved OpenMP load balance in src/soe/ParallelPrimeSieve.cpp.
 5. Improved code readability of EratSmall.cpp, EratMedium.cpp,
    EratBig.cpp, SieveOfEratosthenes.cpp and others.
 6. The Makefile now automatically detects the CPU's L1 data cache
    size on Unix-like OSes (Linux, Mac OS X).
 7. Renamed ./docs to ./doc
 8. Revised README.txt, added 7. Motivation.
 9. Updated INSTALL, LIBPRIMESIEVE and USAGE_EXAMPLES.
10. Added version #defines to PrimeSieve.h, e.g. for this release:

    #define PRIMESIEVE_VERSION       "3.8"
    #define PRIMESIEVE_MAJOR_VERSION  3
    #define PRIMESIEVE_MINOR_VERSION  8
    #define PRIMESIEVE_YEAR           2012

Changes in version 3.7, 31/05/2012
==================================

 1. More aggressive inlining (*-inline.h), up to 10 percent faster
    prime number generation, up to 20 percent faster initialization.
 2. Reduced header file dependencies, libprimesieve now only depends on
    PrimeSieve.h or ParallelPrimeSieve.h (and PrimeSieve.h).
 3. Updated Makefiles.
 4. New template imath.h functions: isqrt(), ilog2(), isPowerOf2(), ...
    isqrt() has been rewritten using Newton's algorithm in order to
    avoid rounding errors of (int)sqrt((double)n) if n > 10^15.
 5. New internal SieveOfEratosthenes::getNextPrime(...) member function.
 6. Replaced old C-style comments with C++ comments.
 7. Updated ExpressionParser to version 2.2.
 8. uin32_t has been replaced by uint_t in src/soe/*.

src/PrimeSieve.cpp
///
/// @file   PrimeSieve.cpp
/// @brief  PrimeSieve is a high level class that manages prime
///         sieving. It is used for printing and counting primes
///         and for computing the nth prime.
///
/// Copyright (C) 2023 Kim Walisch, <kim.walisch@gmail.com>
///
/// This file is distributed under the BSD License. See the COPYING
/// file in the top level directory.
///

#include <primesieve/forward.hpp>
#include <primesieve/PrimeSieve.hpp>
#include <primesieve/ParallelSieve.hpp>
#include <primesieve/CountPrintPrimes.hpp>
#include <primesieve/pmath.hpp>
#include <primesieve/Vector.hpp>
#include <primesieve/PreSieve.hpp>

#include <stdint.h>
#include <algorithm>
#include <chrono>
#include <iostream>

namespace {

struct SmallPrime
{
  uint64_t first;
  uint64_t last;
  int index;
  const char* str;
};

const primesieve::Array<SmallPrime, 8> smallPrimes
{{
  { 2,  2, 0, "2" },
  { 3,  3, 0, "3" },
  { 5,  5, 0, "5" },
  { 3,  5, 1, "(3, 5)" },
  { 5,  7, 1, "(5, 7)" },
  { 5, 11, 2, "(5, 7, 11)" },
  { 5, 13, 3, "(5, 7, 11, 13)" },
  { 5, 17, 4, "(5, 7, 11, 13, 17)" }
}};

} // namespace

namespace primesieve {

PrimeSieve::PrimeSieve()
{
  int sieveSize = get_sieve_size();
  setSieveSize(sieveSize);
}

/// Used for multi-threading
PrimeSieve::PrimeSieve(ParallelSieve* parent) :
  flags_(parent->flags_),
  sieveSize_(parent->sieveSize_),
  parent_(parent)
{ }

void PrimeSieve::reset()
{
  counts_.fill(0);
  percent_ = -1.0;
  seconds_ = 0.0;
  sievedDistance_ = 0;
}

bool PrimeSieve::isFlag(int flag) const
{
  return (flags_ & flag) == flag;
}

bool PrimeSieve::isFlag(int first, int last) const
{
  int mask = (last * 2) - first;
  return (flags_ & mask) != 0;
}

bool PrimeSieve::isCountPrimes() const
{
  return isFlag(COUNT_PRIMES);
}

bool PrimeSieve::isPrintPrimes() const
{
  return isFlag(PRINT_PRIMES);
}

bool PrimeSieve::isPrint() const
{
  return isFlag(PRINT_PRIMES, PRINT_SEXTUPLETS);
}

bool PrimeSieve::isCountkTuplets() const
{
  return isFlag(COUNT_TWINS, COUNT_SEXTUPLETS);
}

bool PrimeSieve::isPrintkTuplets() const
{
  return isFlag(PRINT_TWINS, PRINT_SEXTUPLETS);
}

bool PrimeSieve::isStatus() const
{
  return isFlag(PRINT_STATUS);
}

bool PrimeSieve::isCount(int i) const
{
  return isFlag(COUNT_PRIMES << i);
}

bool PrimeSieve::isPrint(int i) const
{
  return isFlag(PRINT_PRIMES << i);
}

uint64_t PrimeSieve::getStart() const
{
  return start_;
}

uint64_t PrimeSieve::getStop() const
{
  return stop_;
}

uint64_t PrimeSieve::getDistance() const
{
  if (start_ <= stop_)
    return stop_ - start_;
  else
    return 0;
}

uint64_t PrimeSieve::getCount(int i) const
{
  return counts_[i];
}

counts_t& PrimeSieve::getCounts()
{
  return counts_;
}

int PrimeSieve::getSieveSize() const
{
  return sieveSize_;
}

double PrimeSieve::getSeconds() const
{
  return seconds_;
}

PreSieve& PrimeSieve::getPreSieve()
{
  return preSieve_;
}

void PrimeSieve::setFlags(int flags)
{
  flags_ = flags;
}

void PrimeSieve::addFlags(int flags)
{
  flags_ |= flags;
}

void PrimeSieve::setStart(uint64_t start)
{
  start_ = start;
}

void PrimeSieve::setStop(uint64_t stop)
{
  stop_ = stop;
}

/// Set the size of the sieve array in KiB (kibibyte)
void PrimeSieve::setSieveSize(int sieveSize)
{
  sieveSize_ = inBetween(16, sieveSize, 8192);
}

void PrimeSieve::setStatus(double percent)
{
  if (!parent_)
  {
    auto old = percent_;
    percent_ = percent;
    if (isFlag(PRINT_STATUS))
      printStatus(old, percent_);
  }
}

void PrimeSieve::updateStatus(uint64_t dist)
{
  if (parent_)
  {
    // This is a worker thread, so we need
    // to send the update status request
    // to the parent object which handles
    // thread synchronization.
    updateDistance_ += dist;
    if (parent_->tryUpdateStatus(updateDistance_))
      updateDistance_ = 0;
  }
  else
  {
    sievedDistance_ += dist;
    double percent = 100;
    if (getDistance() > 0)
      percent = sievedDistance_ * 100.0 / getDistance();
    auto old = percent_;
    percent_ = std::min(percent, 100.0);
    if (isFlag(PRINT_STATUS))
      printStatus(old, percent_);
  }
}

void PrimeSieve::printStatus(double old, double current)
{
  int percent = (int) current;
  if (percent > (int) old)
  {
    std::cout << '\r' << percent << '%' << std::flush;
    if (percent == 100)
      std::cout << '\n';
  }
}

/// Process small primes <= 5 and small k-tuplets <= 17
void PrimeSieve::processSmallPrimes()
{
  for (auto& p : smallPrimes)
  {
    if (p.first >= start_ && p.last <= stop_)
    {
      if (isCount(p.index))
        counts_[p.index]++;
      if (isPrint(p.index))
        std::cout << p.str << '\n';
    }
  }
}

uint64_t PrimeSieve::countPrimes(uint64_t start, uint64_t stop)
{
  sieve(start, stop, COUNT_PRIMES);
  return getCount(0);
}

void PrimeSieve::sieve(uint64_t start, uint64_t stop)
{
  setStart(start);
  setStop(stop);
  sieve();
}

void PrimeSieve::sieve(uint64_t start, uint64_t stop, int flags)
{
  setStart(start);
  setStop(stop);
  setFlags(flags);
  sieve();
}

/// Sieve the primes and prime k-tuplets (twin primes,
/// prime triplets, ...) in [start, stop].
///
void PrimeSieve::sieve()
{
  reset();

  if (start_ > stop_)
    return;

  setStatus(0);
  auto t1 = std::chrono::system_clock::now();

  if (start_ <= 5)
    processSmallPrimes();

  if (stop_ >= 7)
  {
    CountPrintPrimes countPrintPrimes(*this);
    countPrintPrimes.sieve();
  }

  auto t2 = std::chrono::system_clock::now();
  std::chrono::duration<double> seconds = t2 - t1;
  seconds_ = seconds.count();
  setStatus(100);
}

} // namespace
